lib/puppet/x509/cert_provider.rb

```diff
@@ -0,0 +1,186 @@
+    @certdir = certdir
+    @requestdir = requestdir
+  end
+
+  # Save `certs` to the configured `capath`.
+  #
+  # @param certs [Array<OpenSSL::X509::Certificate>] Array of CA certs to save
+  # @api private
+  def save_cacerts(certs)
+    save_pem(certs.map(&:to_pem).join, @capath)
+  end
+
+  # Load CA certs from the configured `capath`.
+  #
+  # @return (see #load_cacerts_from_pem)
+  # @raise (see #load_cacerts_from_pem)
+  # @api private
+  def load_cacerts
+    pem = load_pem(@capath)
+    pem ? load_cacerts_from_pem(pem) : nil
+  end
+
+  # Load PEM encoded CA certificates.
+  #
+  # @param pem [String] PEM encoded certificate(s)
+  # @return [Array<OpenSSL::X509::Certificate>] Array of CA certs
+  # @raise [OpenSSL::X509::CertificateError] The `pem` text does not contain a valid cert
+  # @api private
+  def load_cacerts_from_pem(pem)
+    pem.scan(CERT_DELIMITERS).map do |text|
+      OpenSSL::X509::Certificate.new(text)
+    end
+  end
+
+  # Save `crls` to the configured `crlpath`.
+  #
+  # @param crls [Array<OpenSSL::X509::CRL>] Array of CRLs to save
+  # @api private
+  def save_crls(crls)
+    save_pem(crls.map(&:to_pem).join, @crlpath)
+  end
+
+  # Load CRLs from the configured `crlpath` path.
+  #
+  # @return (see #load_crls_from_pem)
+  # @raise (see #load_crls_from_pem)
+  # @api private
+  def load_crls
+    pem = load_pem(@crlpath)
+    pem ? load_crls_from_pem(pem) : nil
+  end
+
+  # Load PEM encoded CRL(s).
+  #
+  # @param pem [String] PEM encoded CRL(s)
+  # @return [Array<OpenSSL::X509::CRL>] Array of CRLs
+  # @raise [OpenSSL::X509::CRLError] The `pem` text does not contain a valid CRL
+  # @api private
+  def load_crls_from_pem(pem)
+    pem.scan(CRL_DELIMITERS).map do |text|
+      OpenSSL::X509::CRL.new(text)
+    end
+  end
+
+  # Save named private key in the configured `privatekeydir`. For
+  # historical reasons, names are case insensitive.
+  #
+  # @param name [String] The private key identity
+  # @param key [OpenSSL::PKey::RSA] private key
+  # @api private
+  def save_private_key(name, key)
+    path = to_path(@privatekeydir, name)
+    save_pem(key.to_pem, path)
+  end
+
+  # Load a private key from the configured `privatekeydir`. For
+  # historical reasons, names are case-insensitive.
+  #
+  # @param name [String] The private key identity
+  # @return (see #load_private_key_from_pem)
+  # @raise (see #load_private_key_from_pem)
+  # @api private
+  def load_private_key(name)
+    path = to_path(@privatekeydir, name)
+    pem = load_pem(path)
+    pem ? load_private_key_from_pem(pem) : nil
+  end
+
+  # Load a PEM encoded private key.
+  #
+  # @param pem [String] PEM encoded private key
+  # @return [OpenSSL::PKey::RSA] The private key
+  # @raise [OpenSSL::PKey::RSAError] The `pem` text does not contain a valid key
+  # @api private
+  def load_private_key_from_pem(pem)
+    # set a non-nil passphrase to ensure openssl doesn't prompt
+    OpenSSL::PKey::RSA.new(pem, '')
```

Comment:
what happens if we do have an encrypted key? Is the user just out of luck?

Reply:
Yep, and that's the current behavior of puppet (if you point it at an encrypted private key, but the `Puppet[:passfile]` file doesn't exist.

Reply:
I also filed https://tickets.puppetlabs.com/browse/PUP-9466 to readd support for encrypted private keys. Current support is dubious, and puppetserver doesn't support it (when agent and server on the same host).